From e80082bee6ee3ec0963d3ae640c312312ba92554 Mon Sep 17 00:00:00 2001
From: Lin Yujun <linyujun809@huawei.com>
Date: Wed, 22 Mar 2023 17:31:00 +0800
Subject: [PATCH 224/921] [Huawei] arm64: Optimize the comparison of unsigned
 expressions to avoid compiling error v2

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188396

--------------------------------

3d0714971cb ("[Huawei] arm64: Optimize the comparison of unsigned
expressions to avoid compiling error") uses functional macros to fix
RTOS compiling error, which in turn introduces CodeCheck warning.

In addition, the modification method of the preceding patch returns
true when "min > max", which is different from the original function.

Therefore, the function macro is not used. Instead, val is forcibly
converted to int to solve the RTOS compiling error.

Signed-off-by: Lin Yujun <linyujun809@huawei.com>
---
 arch/arm64/include/asm/cpufeature.h | 18 ++++++------------
 1 file changed, 6 insertions(+), 12 deletions(-)

diff --git a/arch/arm64/include/asm/cpufeature.h b/arch/arm64/include/asm/cpufeature.h
index f78f7f5d53e5..cfa6eb20cacb 100644
--- a/arch/arm64/include/asm/cpufeature.h
+++ b/arch/arm64/include/asm/cpufeature.h
@@ -53,9 +53,6 @@ enum ftr_type {
 #define FTR_VISIBLE_IF_IS_ENABLED(config)		\
 	(IS_ENABLED(config) ? FTR_VISIBLE : FTR_HIDDEN)
 
-#define IN_RANGE_INCLUSIVE(val, min, max)		\
-	(((val) - (min)) < ((val) - (max) - 1))
-
 struct arm64_ftr_bits {
 	bool		sign;	/* Value is signed ? */
 	bool		visible;
@@ -653,9 +650,8 @@ static inline bool system_supports_4kb_granule(void)
 	val = cpuid_feature_extract_unsigned_field(mmfr0,
 						ID_AA64MMFR0_TGRAN4_SHIFT);
 
-	return IN_RANGE_INCLUSIVE(val,
-		ID_AA64MMFR0_TGRAN4_SUPPORTED_MIN,
-		ID_AA64MMFR0_TGRAN4_SUPPORTED_MAX);
+	return ((int)val >= ID_AA64MMFR0_TGRAN4_SUPPORTED_MIN) &&
+		(val <= ID_AA64MMFR0_TGRAN4_SUPPORTED_MAX);
 }
 
 static inline bool system_supports_64kb_granule(void)
@@ -667,9 +663,8 @@ static inline bool system_supports_64kb_granule(void)
 	val = cpuid_feature_extract_unsigned_field(mmfr0,
 						ID_AA64MMFR0_TGRAN64_SHIFT);
 
-	return IN_RANGE_INCLUSIVE(val,
-		ID_AA64MMFR0_TGRAN64_SUPPORTED_MIN,
-		ID_AA64MMFR0_TGRAN64_SUPPORTED_MAX);
+	return ((int)val >= ID_AA64MMFR0_TGRAN64_SUPPORTED_MIN) &&
+		(val <= ID_AA64MMFR0_TGRAN64_SUPPORTED_MAX);
 }
 
 static inline bool system_supports_16kb_granule(void)
@@ -681,9 +676,8 @@ static inline bool system_supports_16kb_granule(void)
 	val = cpuid_feature_extract_unsigned_field(mmfr0,
 						ID_AA64MMFR0_TGRAN16_SHIFT);
 
-	return IN_RANGE_INCLUSIVE(val,
-		ID_AA64MMFR0_TGRAN16_SUPPORTED_MIN,
-		ID_AA64MMFR0_TGRAN16_SUPPORTED_MAX);
+	return ((int)val >= ID_AA64MMFR0_TGRAN16_SUPPORTED_MIN) &&
+		(val <= ID_AA64MMFR0_TGRAN16_SUPPORTED_MAX);
 }
 
 static inline bool system_supports_mixed_endian_el0(void)
-- 
2.17.1

